### ISSUE TITLE
some questions about geomopt

### ISSUE THREAD

**Issue by jamcas** [pyscf/pyscf#431]
Hi, i want to do geometric optimination by casscf, but there are some problems hinderded me:

1.How to read orbitals from a converged calculation? I followed this example :(https://github.com/pyscf/pyscf/blob/dev/examples/geomopt/12-mcscf_excited_states.py), but failed:
```
from pyscf import gto
from pyscf import scf, mcscf

mol = gto.Mole()
mol.atom="N; N 1, 1.1"
mol.basis= "6-31g"
mol.build()

mf = scf.RHF(mol).run()

mc = mcscf.CASSCF(mf, 4,4)
mc.state_average_([0.5, 0.5])
mc.run()

mo = mc.mo_coeff

mc.state_specific_(2)
excited_grad = mc.nuc_grad_method().as_scanner()
mol1 = excited_grad.optimizer().kernel(mo)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'CASSCF_GradScanner' object has no attribute 'optimizer'
```
2.Why do we need to performe a dft calculation  in each optimization iteration instead of directly using the previous converged casscf calculation as the guess?

3.How to print the large CI componets after each iteration in geomopt? As we know,  when optiming the excited states, states may flip, so i need to check the state after each iteration in order to take some measures.

**Comment by jamesETsmith** [comment]
Hi @jamcas,

1) To read MOs after an optimization you can use chkfiles which will keep track of the orbitals throughout the optimization. See this for the usage of the chkfile utilities in PySCF:
https://github.com/pyscf/pyscf/blob/41ac598523722eeeb4c39e1723a6bf80a7c889cd/examples/mcscf/13-load_chkfile.py#L50-L55

2) You do not need to perform a DFT from scratch at every iteration. See here for the PySCFEngine used by the geomeTRIC solver
https://github.com/pyscf/pyscf/blob/41ac598523722eeeb4c39e1723a6bf80a7c889cd/pyscf/geomopt/geometric_solver.py#L113-L118

(fyi the example you provided is doing CASSCF not DFT).

3) You will probably need to modify the CASSCF scanner object to print out the configurations at each step using code similar to this example:
https://github.com/pyscf/pyscf/blob/a5c890f4ed5bd8cc8c2e417ff885d0b99062150d/examples/fci/11-large_ci.py#L27-L38
For more details about printing CI coefficients and configurations see #145 
@sunqm may be able to provide a more elegant solution here.

**Comment by jamcas** [comment]
@jamesETsmith, thanks so much for your kind reply, it helps me a lot.

**Comment by sunqm** [comment]
@jamcas 
1.  You can assign orbitals to mc.mo_coeff as initial guess
```
mc.state_specific_(2)
mc.mo_coeff = mo
```
.optimizer method is supported only by the dev branch.

Some recent updates in dev branch fixed a few problems when calling mc.state_specific after mc.state_average

**Comment by sunqm** [comment]
3. A callback function can be passed to geometry optimizer. It's called after each iteration. To analyze mcscf wfn, you can set callback as a quick solution
```
excited_grad = mc.nuc_grad_method().as_scanner()
opt = excited_grad.optimizer()
def cb(envs):
    mc = envs['g_scanner'].base
    mc.analyze(verbose=4)
opt.callback = cb
mol1 = opt.kernel()
```

I've added relevant examples in  https://github.com/pyscf/pyscf/blob/master/examples/geomopt/20-callback.py

**Comment by jamcas** [comment]
@sunqm, that's great! I also have a question about the geometric optimination for excited state .What should I do If i want to optimize the 5th state among 8 roots? Shoud I do a 8 roots state-average calculation firstly, then read it orbital and set mc.state_specific_(5)? Or is there any other way?

**Comment by sunqm** [comment]
@jamcas I do not understand the problem. Is the entire geometry optimization targeting at only the 5th state? If so, why a 8-root state-average is needed? I think it can be done by optimizing the state-specific calculation from beginning.

**Comment by jamcas** [comment]
@sunqm In my case, The energy of the fifth, sixth and seventh state is particularly close, and their difference is around 1kcal/mol, so when I go directly to optimize the fifth state, it may slip away, I might get the sixth or seventh state.

**Comment by sunqm** [comment]
In other words, it's an optimization of a specific state in the state-averaged CASSCF states. The CASSCF gradient code cannot be used here. It is because the energy of the 5th state in the SA-CASSCF states is not variational wrt the orbital coefficients. The state-specific case is closed to the  CASCI calculation. But the pyscf CASCI gradient code cannot be used either because the CASCI gradient code only supports HF orbitals. The orbital from SA-CASSCF is not subject to this case.

**Comment by jamcas** [comment]
Yes, you are right, so does the current version of pyscf not support such calculations?

**Comment by MatthewRHermes** [comment]
@jamcas I wrote an extension to PySCF that does this kind of gradient calculation in github.com/MatthewRHermes/mrh.git.  Use the 'dev' branch and look in the examples/sacasscf_gradients folder.  You'll also have to use my fork of PySCF at github.com/MatthewRHermes/pyscf.git with the branch 'sapdft_geomopt_compatible' for the moment (until I get around to fixing a small convention difference w/ PySCF in my 'mrh' code), but I just merged that with the current dev branch so it shouldn't be any different other than 2 or 3 lines.
